packages/mobx-state-tree/src/types/complex-types/map.ts

```diff
@@ -42,43 +42,58 @@ import {
     TypeFlags,
     Union,
     EMPTY_OBJECT,
-    OptionalProperty
+    OptionalProperty,
+    ExtractC,
+    ExtractS,
+    ExtractT
 } from "../../internal"
 
-export interface IMapType<C, S, T>
-    extends IComplexType<IKeyValueMap<C> | undefined, IKeyValueMap<S>, IMSTMap<C, S, T>>,
+export interface IMapType<IT extends IAnyType>
+    extends IComplexType<
+            IKeyValueMap<ExtractC<IT>> | undefined,
+            IKeyValueMap<ExtractS<IT>>,
+            IMSTMap<IT>
+        >,
         OptionalProperty {}
 
-export interface IMSTMap<C, S, T> {
+export interface IMSTMap<IT extends IAnyType> {
     // bases on ObservableMap, but fine tuned to the auto snapshot conversion of MST
 
     clear(): void
     delete(key: string): boolean
-    forEach(callbackfn: (value: T, key: string, map: IMSTMap<C, S, T>) => void, thisArg?: any): void
-    get(key: string): T | undefined
+    forEach(
+        callbackfn: (value: ExtractT<IT>, key: string, map: IMSTMap<IT>) => void,
+        thisArg?: any
+    ): void
+    get(key: string): ExtractT<IT> | undefined
     has(key: string): boolean
-    set(key: string, value: C | S | T): this
+    set(key: string, value: ExtractC<IT> | ExtractS<IT> | ExtractT<IT>): this
```

Comment:
We could introduce `ExtractCST<IT>` to simplify?

Reply:
will do